spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java

```diff
@@ -386,42 +390,89 @@ private void createDataProcessorStatement(VariableExpression variable, Expressio
   }
 
   private void rewriteBinaryTableLikeParameterization(ListIterator<Statement> stats) throws InvalidSpecCompileException {
-    rewriteTableLikeParameterization(stats, result -> {
+    rewriteTableLikeParameterization(stats, () -> {
       Statement stat = stats.next();
       BinaryExpression orExpr = getOrExpression(stat);
       if (orExpr == null) {
         stats.previous();
-        return true;
+        return null;
       }
-      splitRow(orExpr, result);
-      return false;
+      List<Expression> row = new ArrayList<>();
+      splitRow(orExpr, row);
+      return new DataTableRow(row, DATA_TABLE_CROSS_PRODUCT_LABEL.equals(stat.getStatementLabel()));
     });
   }
 
   private void rewriteExpressionTableLikeParameterization(ListIterator<Statement> stats) throws InvalidSpecCompileException {
-    rewriteTableLikeParameterization(stats, result -> {
+    rewriteTableLikeParameterization(stats, () -> {
+      boolean headerRow;
+      if (stats.hasNext()) {
+        headerRow = DATA_TABLE_CROSS_PRODUCT_LABEL.equals(stats.next().getStatementLabel());
+        stats.previous();
+      } else {
+        headerRow = false;
+      }
+
       List<Expression> row = getExpressionChain(stats);
       if (row.size() <= 1) {
         // rewind
         row.forEach(expression -> stats.previous());
-        return true;
+        return null;
       }
-      result.addAll(row);
-      return false;
+      return new DataTableRow(row, headerRow);
     });
   }
 
-  private void rewriteTableLikeParameterization(ListIterator<Statement> stats, Function<List<Expression>, Boolean> rowExtractor) throws InvalidSpecCompileException {
+  private void rewriteTableLikeParameterization(
+    ListIterator<Statement> stats,
+    ThrowingSupplier<DataTableRow, InvalidSpecCompileException> rowExtractor) throws InvalidSpecCompileException {
+
+    List<List<Expression>> baseRows = null;
     LinkedList<List<Expression>> rows = new LinkedList<>();
+    int baseRowSize = 0;
+    ASTNode lastHeaderLocation = null;
 
     while (stats.hasNext()) {
-      List<Expression> row = new ArrayList<>();
-      if (rowExtractor.apply(row)) {
+      DataTableRow row = rowExtractor.get();
+      if (row == null) {
         break;
       }
-      if (rows.size() > 0 && rows.getLast().size() != row.size())
-        throw new InvalidSpecCompileException(row.get(0), String.format("Row in data table has wrong number of elements (%s instead of %s)", row.size(), rows.getLast().size()));
-      rows.add(row);
+
+      if (lastHeaderLocation == null) {
+        lastHeaderLocation = row.expressions.get(0);
+      }
+
+      if (row.isHeader) {
```

Comment:
Can we extract these two branches to methods

Reply:
Actually the `if` branch has four outputs, so not so easy except if I introduce some kind of state object.